### PR TITLE
8333749: Consolidate ConstantDesc conversion in java.base

### PR DESCRIPTION
**Pull request by liach**
In java.base, especially in bytecode generators, we have many different methods converting known valid Class and MethodType into ClassDesc and MethodTypeDesc. These conversions should be consolidated into the same utility method for the ease of future maintenance.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333749](https://bugs.openjdk.org/browse/JDK-8333749): Consolidate ConstantDesc conversion in java.base (**Enhancement** - P4)

### Reviewers
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**)
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to [e9ddc398](https://git.openjdk.org/jdk/pull/19585/files/e9ddc3987d4e730b5be0ac978821683943fa5726)

### Contributors
 * Claes Redestad `<redestad@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19585/head:pull/19585` \
`$ git checkout pull/19585`

Update a local copy of the PR: \
`$ git checkout pull/19585` \
`$ git pull https://git.openjdk.org/jdk.git pull/19585/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19585`

View PR using the GUI difftool: \
`$ git pr show -t 19585`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19585.diff">https://git.openjdk.org/jdk/pull/19585.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19585#issuecomment-2153166381)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java [hunk 1]
```diff
@@ -466,9 +468,9 @@ private void doBindings(List<Binding> bindings) {
                 case Copy copy                   -> emitCopyBuffer(copy);
                 case Allocate allocate           -> emitAllocBuffer(allocate);
                 case BoxAddress boxAddress       -> emitBoxAddress(boxAddress);
-                case SegmentBase unused          -> emitSegmentBase();
+                case SegmentBase _               -> emitSegmentBase();
```
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:471]: Seem a bit too far detached from the changes at hand for a drive-by code cleanup?

**Review by cl4es — APPROVED** [review]
Looks good to me. Probably should get someone to OK changes in foreign/abi (@JornVernee perhaps?).

There are some pre-existing places where we call `ReferenceClassDescImpl.ofValidated` directly that could probably be switched over to `ConstantUtils.classDesc` for slightly nicer code. Or - if it matters - add a `referenceClassDesc` which avoids the `type.isPrimitive` test.

File: src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java [hunk 2]
```diff
@@ -464,10 +466,8 @@ record Element(Label target, Label next, Object caseLabel) { }
                             // Object o = ...
                             // o instanceof Wrapped(float)
                             cb.aload(SELECTOR_OBJ);
-                            cb.instanceOf(Wrapper.forBasicType(classLabel)
-                                    .wrapperType()
-                                    .describeConstable()
-                                    .orElseThrow());
+                            cb.instanceOf(classDesc(Wrapper.forBasicType(classLabel)
```
[the remark this answers is not in the document]
Comment (liach) [reply src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java:469]: Feel free, or you can get your patch merged first and then I base off yours if yours is ready.

File: src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java (hunk 1 above)
Comment (liach) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:471]: My uncontrollable urges to fix these minor IDE warnings... Please allow these fixes, as I have similar changes in ProxyGenerator as well.

File: src/java.base/share/classes/jdk/internal/constant/ConstantUtils.java [hunk 3]
```diff
@@ -47,6 +49,59 @@ public final class ConstantUtils {
...
+        }
+        return ReferenceClassDescImpl.ofValidated(type.descriptorString());
+    }
+
+    /**
+     * Creates a MethodTypeDesc from a MethodType object, requires that
+     * the type can be described nominally, i.e. all of its return
+     * type and parameter types can be described nominally.
+     */
+    public static MethodTypeDesc methodDesc(MethodType type) {
```
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/constant/ConstantUtils.java:76]: Please name these methods `methodTypeDesc`, since we also have the `Method` type.

**Comment by liach** [comment]
> There are some pre-existing places where we call `ReferenceClassDescImpl.ofValidated` directly that could probably be switched over to `ConstantUtils.classDesc` for slightly nicer code. Or - if it matters - add a `referenceClassDesc` which avoids the `type.isPrimitive` test.

I think the main advantage of `ofValidated` is that it avoids descriptor string computation from class objects and should be kept in performance-sensitive code.

`referenceClassDesc` would be feasible for less performance-demanding code's initialization and known safe conversions (a few APIs in CF API expects non-primitive CDs already). But should its implementation be `ofValidated` or `ofValidatedBinaryName` (same question for ClassDesc)? We used `ofValidatedBinaryName` here instead of a `toClassDesc` (which calls `ofValidated` via `descriptorString`)

File: src/java.base/share/classes/jdk/internal/constant/ConstantUtils.java (hunk 3 above)
Comment (liach) [reply src/java.base/share/classes/jdk/internal/constant/ConstantUtils.java]: Thanks for the suggestion. Done.

File: src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cl4es) [reply src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java]: Created a PR you could fold into this, if you'd like: https://github.com/liachmodded/jdk/pull/1

**Comment by ExE-Boss** [comment]
This could probably also update `Class​::describeConstable()` to use `ReferenceClassDesc​::ofValidated​(String)` for non‑hidden reference types:
https://github.com/openjdk/jdk/blob/40b2fbd8207404961d3d23375b288cceafc3f902/src/java.base/share/classes/java/lang/Class.java#L4709-L4713

**Review by JornVernee — APPROVED** [review]

**Review by cl4es — APPROVED** [review]
Good incremental improvements.

**Comment by liach** [comment]
Thanks for the reviews!

/integrate

**Comment by ExE-Boss** [comment]
`ClassDesc​::of​(String)`, `ClassDesc​::of​(String, String)`, and `ClassDesc​::ofInternalName​(String)` can probably call `ReferenceClassDescImpl​::ofValidated​(String)` directly as those methods already perform validation before forwarding to `ClassDesc​::ofDescriptor​(String)`.

**Comment by liach** [comment]
No: they are missing checks against initial, trailing, or consecutive . or / separators. We can do that in a future patch, with proper test additions.

**Comment by liach** [comment]
Thanks for the reviews!

/integrate
